### ISSUE TITLE
fd_readdir not behaving according to the specification

### ISSUE THREAD

**Issue by mischnic** [wasmerio/wasmer-js#164]
I think the fs_readdir implementation is wrong:

### Current Behaviour

According to
https://github.com/WebAssembly/WASI/blob/1d1358a/phases/old/snapshot_0/docs/wasi_unstable.md#fd_readdir:

> This function fills the output buffer as much as possible, potentially truncating the last directory entry. This allows the caller to grow its read buffer size in case it's too small to fit a single large directory entry, or skip the oversized directory entry.

> `bufused`
> The number of bytes stored in the read buffer. **If less than the size of the read buffer, the end of the directory has been reached.**

But `@wasmer/wasi` sets `bufused` to a smaller value although there are still files in the directory that weren't enumerated:
https://github.com/wasmerio/wasmer-js/blob/4cd68ac620d51cb220d7d59d58a2b5f964a24344/packages/wasi/src/index.ts#L784-L787
https://github.com/wasmerio/wasmer-js/blob/4cd68ac620d51cb220d7d59d58a2b5f964a24344/packages/wasi/src/index.ts#L829-L830

### Expected Behaviour

fd_readdir shouldn't stop early if the buffer isn't large enough for the entry + filename but instead blindly write into the buffer until it's full (even if that means some data of the current entry is discarded):
- if the next entry doesn't fit in the buffer: break and set bufused to bufsize
- else: write the entry, then:
	- if the filename doesn't fit in the buffer: break and set bufused to bufsize
	- else: continue!

That would make also wasi-libc behave correctly: https://github.com/CraneStation/wasi-libc/blob/f645f498dfbbbc00a7a97874d33082d3605c3f21/libc-bottom-half/cloudlibc/src/libc/dirent/readdir.c#L60-L66

### Reproduction

```c
#include <dirent.h>
#include <stdio.h>
#include <string.h>

void listdir(const char *name, int indent) {
	DIR *dir;
	struct dirent *entry;

	printf("dirent: %lu\n", sizeof(__wasi_dirent_t));

	if (!(dir = opendir(name)))
		return;

	while ((entry = readdir(dir)) != NULL) {
		printf("%*s- %s\n", indent, "", entry->d_name);
	}
	closedir(dir);
}

int main(void) {
	listdir("/absolute/path/to/readdir_test", 0);
	return 0;
}
```

```sh
# compile
clang --target=wasm32-unknown-wasi --sysroot ..../wasi-libc/sysroot test.c

# Setup folder which will fill wasi-libc's default buffer size of 4096:
mkdir readdir_test
touch readdir_test/aaaaaaaaaaaaaaaaa{0..100}

# run
wasmer-js run a.out --dir=/
```

The output is (files 90-100 aren't printed):

```
...
- aaaaaaaaaaaaaaaaaa084
- aaaaaaaaaaaaaaaaaa085
- aaaaaaaaaaaaaaaaaa086
- aaaaaaaaaaaaaaaaaa087
- aaaaaaaaaaaaaaaaaa088
- aaaaaaaaaaaaaaaaaa089
```

**Comment by torch2424** [comment]
@mischnic Thank you soooo much for such a detailed issue!

cc @syrusakbary as they have been doing a lot of the Wasi polyfill work here.

If they don't get the chance to get to it, I'll try to when I get the chance! 😄

**Comment by syrusakbary** [comment]
We just published a new version of wasmer-js `0.5.1` that should have this issue fixed.
